A Step Towards Improving Map Layer Display

authenticate opened the pull request:
Hi,

I want to preface this by saying I know you had said you worked on this some.  So, do not feel obligated or required to use this code.  I had some free time and felt like this would be a fun feature to work on.

This pull request adds a class which allows off screen rendering and/or render to texture functionality.  In addition, it allows the rendered texture to be used in subsequent drawing operations.

This pull request DOES NOT actually solve the map layer display issue.  It provides the necessary mechanisms and systems to solve the problem.  (It composites the map into an off screen texture.  Then, it draws the final, composited map onto the screen.)  However, it does not yet draw the map at its native resolution nor scale up the native, composited map to its final resolution.

There are also a couple of minor changes which I will address as comments in the code itself.

Let me know if you have any questions or comments.

Thanks.

authenticate commented on src/engine/video/video.cpp:
```diff
@@ -126,91 +129,16 @@ VideoEngine::VideoEngine():
         _fps_samples[sample] = 0;
 }
 
-void VideoEngine::_UpdateFPS()
```
The header file declared these functions near the bottom.  I just moved these functions to the bottom of the implementation file.  Nothing should have changed within the functions themselves.

Bertram25 replied:
k

authenticate commented on src/engine/video/video.cpp:
```diff
@@ -404,17 +351,11 @@ void VideoEngine::SetDrawFlags(int32_t first_flag, ...)
 
 void VideoEngine::Clear()
 {
-    //! \todo glClearColor is a state change operation. It should only be called when the clear color changes
```
I moved the 'glClearColor' call to an initialization function.  The clear color was never changed at run-time anyways.

In addition, there was a subtle viewport bug where the cached viewport variables would become out of sync with the viewport itself.

So, the only 'glViewport' call should now be in the 'SetViewport' function.  I also removed the viewport update from the 'Clear' function.

Bertram25 replied:
Sounds good to me.

authenticate commented on src/modes/map/map_mode.cpp:
```diff
@@ -1073,6 +1073,19 @@ void MapMode::_DrawMapLayers()
 {
     VideoManager->PushState();
     VideoManager->SetStandardCoordSys();
+
```
Somewhere in here is where I imagine all the heavy lifting will eventually happen.

Ideally, I think everything (Not just map tiles; the map objects and the player character too.) should be drawn at the native scale of the map tiles which I think is 32x32 pixels.  Then, scaled up at the end.

I did some initial tinkering to try and diagnose what would need to happen.  However, I ran into some issues.  I could never get the map tiles and all the map objects to align in any other way than they are currently.  Swapping their coordinate systems and scales around seemed pretty tough to do correctly.

Bertram25 replied:
> Ideally, I think everything (Not just map tiles; map objects and player character too.) should be drawn at the native scale of the map tiles which I think is 32x32 pixels. Then, everything can be scaled up at the end.

Yes, exactly. That would in the end make the engine code cleaner, and much more size agnostic in any case.

> I did some initial tinkering to try and diagnose what would need to happen. However, I ran into some issues. I could never get the map tiles and all the map objects to align in any other way than they are currently. Swapping their coordinate systems and scales around seemed pretty tough to do correctly.

I do think that's because of the overall logic computing the tiles location on screen vs the way it is done for sprites. I guess you already saw where tile offset begin and end are set up, and how sprites are placed on screen?

authenticate commented on src/main.cpp:
```diff
@@ -624,11 +592,29 @@ int main(int argc, char *argv[])
         return EXIT_FAILURE;
     }
 
+    // NOTE: Even if the singleton objects do not exist when this function is called, invoking the
```
I moved the contents of the clean up callback function here.

```
SDL_GL_DeleteContext(glcontext);
```

was being called before

```
VideoEngine::SingletonDestroy();
```

This caused the OpenGL context to be destroyed before the clean up its resources actually happened.  So, now the VideoEngine is cleaned up prior to deleting the OpenGL context.

Bertram25 replied:
k, thanks.

Bertram25 commented:
> I want to preface this by saying I know you had said you worked on this some. So, do not feel obligated or required to use this code. I had some free time and felt like this would be a fun feature to work on.

Spare-time apart, yes I worked on it a bit, but I must say I find pretty cool from you to pour some energy in it. :) Will look at this one eagerly.

> This pull request DOES NOT actually solve the map layer display issue. It provides the necessary mechanisms and systems to solve the problem. (It composites the map into an off screen texture. Then, it draws the final, composited map onto the screen.) However, it does not yet draw the map at its native resolution nor scale up the native, composited map to its final resolution.

Understood, what seems to be already achieved is awesome already, so no need to apologize about not finishing it in any case. :)

Will review now, and test.

Bertram25 commented on src/engine/video/gl/gl_render_target.cpp:
```diff
@@ -0,0 +1,292 @@
+    assert(_width > 0);
+    assert(_height > 0);
+
+    bool errors = false;
```
The `errors` member and all the logic around it looks weird to me.
IMHO, as soon as you've got a `glGetError() != GL_NO_ERROR` you should print an error and trigger the end of app before returning if quit app is non-blocking and doesn't actual suffice to stop the game.

IMHO, all in all, you can remove all the `if (errors)` as they are redundant.
WDYT?

authenticate replied:
Yes, I agree.

So, I always try to write functions which never exit early.  They run to completion every time and do not have cases which cause them to short circuit and exit early.

I believe code written in this manner is easier to read and to understand.  (This is debatable and subjective.  It could be totally wrong.) :)

That is why it is written the way it is.  Now, that does not mean it is great or perfect or does not need changed.  However, that is it at least the methodology of what I was trying to achieve.

So, the code should proceed from top to bottom and do basically one thing per 'if' block.  If something goes wrong, the code should log the problem, update the flag, and proceed downwards to the next 'if' block.  However, in this case, it will skip over any remaining tasks and eventually exit.

I think one of the big weakness, which you pointed out, is the caller never knows about any problems (other than the log messages).  So, the code always succeeds, even when it does not actually or fully succeed.

Bertram25 commented on src/engine/video/video.h:
```diff
@@ -332,6 +322,15 @@ class VideoEngine : public vt_utils::Singleton<VideoEngine>
+    void DisableSecondaryRenderTarget();
+
+    //! Draws the secondary render target onto the primary render target.
+    void DrawSecondaryRenderTarget();
```
As this is actually disabling the secondary render target, maybe this should be stated, right?

Bertram25 commented:
Got some warnings when compiling on linux:
```shell
ValyriaTear/src/engine/video/video.cpp: In member function ‘void vt_video::VideoEngine::DrawSecondaryRenderTarget()’:
ValyriaTear/src/engine/video/video.cpp:620:11: warning: unused variable ‘width_screen’ [-Wunused-variable]
     float width_screen = static_cast<float>(vt_video::VideoManager->GetScreenWidth());
           ^
ValyriaTear/src/engine/video/video.cpp:621:11: warning: unused variable ‘height_screen’ [-Wunused-variable]
     float height_screen = static_cast<float>(vt_video::VideoManager->GetScreenHeight());
           ^
```

Bertram25 commented:
Ah, and unfortunately, got a segfault at startup:
```shell
Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x0000000000c27ccd in vt_video::gl::RenderTarget::RenderTarget (this=0x28d7ba0, width=1024, height=768) at ValyriaTear/src/engine/video/gl/gl_render_target.cpp:106
#2  0x0000000000c709b8 in vt_video::VideoEngine::FinalizeInitialization (this=0x14b1080) at ValyriaTear/src/engine/video/video.cpp:198
#3  0x0000000000e7e5f9 in InitializeEngine () at ValyriaTear/src/main.cpp:382
#4  0x0000000000e7f4ad in main (argc=1, argv=0x7fffffffe128) at ValyriaTear/src/main.cpp:502
```

authenticate commented:
I updated the code to remove the warnings.

In addition, I changed the function causing the segfault to a different function.  Let me know if that fixes the segfault or if some other issue happens.  The function I changed it to is more explicit than the original function.  So, hopefully that helps.

Bertram25 commented:
Ok, I'll be on this asap. Thanks a lot for the work done! :D

Bertram25 commented:
K, the code looks good apart for the point we are discussing (See afterward.). I'll test it asap, and merge if everything looks good. :)

I believe code written in this manner is easier to read and to understand. (This is debatable and subjective. It could be totally wrong.) :)

>That is why it is written the way it is. Now, that does not mean it is great or perfect or does not need changed. However, that is it at least the methodology of what I was trying to achieve.
So, the code should proceed from top to bottom and do basically one thing per 'if' block. If something goes wrong, the code should log the problem, update the flag, and proceed downwards to the next 'if' block. However, in this case, it will skip over any remaining tasks and eventually exit.
I think one of the big weakness, which you pointed out, is the caller never knows about any problems (other than the log messages). So, the code always succeeds, even when it does not actually or fully succeed.

If the current code can reach merge state, I'll rework the function part with the errors point, to see how different it is from your point of view, you'll then be able to comment my code. :)

Bertram25 commented:
The game runs. :)

Actually, you should have a game start, try to move up and down in the first map (the upper room), and try to talk to people.
As for me, I identified 2 bugs relative to the proposed tests:
- Sometimes part of the map is not drawn anymore. I suspect it is because of the algorithm detecting whether a map part has to be drawn or not. (Happens at 800x600 everywhere, and at 1064x768 is Flora's Shop)
![screenshot_3](https://cloud.githubusercontent.com/assets/765917/11908160/4bca50a0-a5da-11e5-8065-4b9c5f687b09.png)
![screenshot_4](https://cloud.githubusercontent.com/assets/765917/11908165/580d2b12-a5da-11e5-9ed5-48ae41a7700e.png)
- Sprite dialogue bubble is weird. (at 800x600 only, at 1024x768, most of it is fine but sometimes the bubble disappear suddenly.)
![screenshot_5](https://cloud.githubusercontent.com/assets/765917/11908166/5cf7d82a-a5da-11e5-8ded-58c0652c1c9f.png)
![screenshot_6](https://cloud.githubusercontent.com/assets/765917/11908168/5e7e7f78-a5da-11e5-9e56-60d35940ad93.png)

authenticate commented:
Thanks for the update.  I can readily reproduce the chat bubble issues.  I will look into those first.

The map drawing issue does not seem as drastic on machine.  I can reproduce cases where a single tile along an edge will not draw.  However, I can not reproduce the extreme cases shown above.

However, I'll start investigating these issues and then we can see where the code stands at that point.

authenticate commented:
I pushed up two commits to address some of these issues.

The location of the chat bubbles was in the wrong position because the projection matrix was not updated within the "PopState" function.  I updated the function to correct this.  I probably caused this a while back when the code base shifted to the programmable OpenGL pipeline.  However, it never actually caused a problem until now.

In addition, the chat bubbles appeared black because the texture controller would sometimes not set the correct texture because it believe the texture was already set in OpenGL.  This was caused because the new render target code manipulates textures too.  I removed the optimization in the texture controller to get things working.  This should be fine.  I do not think this will cripple our performance.

Finally, let me know if the map still draws incorrectly.  The projection matrix update may have fixed it.  I am not entirely sure.  It seems to work alright now on my machine.

Thanks again.

Bertram25 commented:
I'll test now. :) thanks for the fixes!

Bertram25 commented:
Tested it thoroughly. Works fine :D Merging!
